streaming/replication/replication.cc

```diff
@@ -0,0 +1,467 @@
+    char Buf[Len];
+    memset(Buf, 0, Len);
+    replication_load_gaps(&RH->host_uuid, Buf, Len);
+    std::vector<TimeRange> TRs = deserializeTimeRanges(Buf, Len);
+
+    /*
+     * Create host
+    */
+    Host *H = new Host(RH);
+    H->setReceiverGaps(TRs);
+    RH->repl_handle = static_cast<replication_handle_t>(H);
+
+    /*
+     * Log info
+     */
+    auto &L = H->getLogger();
+    L.createdHost(TRs);
+}
+
+void replication_delete_host(RRDHOST *RH) {
+    Host *H = static_cast<Host *>(RH->repl_handle);
+    if (!H)
+        return;
+
+    /*
+     * Save receiver gaps to sqlite DB
+     */
+    size_t Len = 8192;
+    char Buf[Len];
+    memset(Buf, 0, Len);
+
+    std::vector<TimeRange> TRs = H->getReceiverGaps();
+    serializeTimeRanges(TRs, Buf, Len);
+    replication_save_gaps(&RH->host_uuid, Buf, Len);
+
+    /*
+     * Log info
+     */
+    auto &L = H->getLogger();
+    L.deletedHost(TRs);
+
+    /*
+     * Delete host
+     */
+    delete H;
+    RH->repl_handle = nullptr;
+}
+
+void replication_thread_start(RRDHOST *RH) {
+    Host *H = static_cast<Host *>(RH->repl_handle);
+    if (!H)
+        return;
+
+    H->startReplicationThread();
+
+    /*
+     * Log info
+     */
+    auto &L = H->getLogger();
+    L.startedReplicationThread();
+}
+
+void replication_thread_stop(RRDHOST *RH) {
+    Host *H = static_cast<Host *>(RH->repl_handle);
+    if (!H)
+        return;
+
+    H->stopReplicationThread();
+
+    /*
+     * Log info
+     */
+    auto &L = H->getLogger();
+    L.stoppedReplicationThread();
+}
+
+void replication_receiver_connect(RRDHOST *RH, char *Buf, size_t Len) {
+    Host *H = static_cast<Host *>(RH->repl_handle);
+    if (!H)
+        return;
+
+    H->receiverConnect();
+    std::vector<TimeRange> TRs = H->getReceiverGaps();
+    serializeTimeRanges(TRs, Buf, Len);
+
+    /*
+     * Log info
+     */
+    auto &L = H->getLogger();
+    L.receiverSentGaps(TRs);
+}
+
+void replication_sender_connect(RRDHOST *RH, const char *Buf, size_t Len) {
+    Host *H = static_cast<Host *>(RH->repl_handle);
+    if (!H)
+        return;
+
+    std::vector<TimeRange> TRs = deserializeTimeRanges(Buf, Len);
+
+    /* Assign the recv'd gaps to the host. The parent sends the gaps
+     * in increasing timestamp order; reverse the vector because
+     * we always pop from the back */
+    std::reverse(TRs.begin(), TRs.end());
+    H->setSenderGaps(TRs);
+
+    /*
+     * Log info
+     */
+    auto &L = H->getLogger();
+    L.senderReceivedGaps(TRs);
+}
+
+bool replication_receiver_fill_gap(RRDHOST *RH, const char *Buf) {
+    char LogPath[FILENAME_MAX + 1];
+    snprintfz(LogPath, FILENAME_MAX, "%s/%s.fg",
+              netdata_configured_cache_dir, rrdhost_hostname(RH));
+    FILE *FP = fopen(LogPath, "a");
+    if (!FP) {
+        error("Could not open log file: %s", LogPath);
+        return false;
+    }
+    size_t BufLen = strlen(Buf);
+    fprintf(FP, "%zu %s\n", BufLen, Buf);
+    fclose(FP);
```

Comment:
@stelfrag This piece of code will log every FILLGAP command received by upstream to `var/netdata/cache/<hostname>.fg`.

If you get a crash within `pluginsd_fillgap_action`, I can check if it's caused by a serialization/deserialization error.

I didn't manage to make the child/parent agents crash during my testing. However, after rebasing on top of master, I'm noticing that not all gaps are filled completely (which should be a separate issue from the crashes that we are trying to investigate now).